subsys/emds/emds.c

```diff
@@ -6,55 +6,353 @@
 
 #include <emds/emds.h>
 
+#include <zephyr.h>
+#include <storage/flash_map.h>
+#include <drivers/flash.h>
+#include <emds/emds_flash.h>
+
 #include <logging/log.h>
 LOG_MODULE_REGISTER(emds, CONFIG_EMDS_LOG_LEVEL);
 
+#define EMDS_STACK_SIZE 500
+#define EMDS_PRIORITY -CONFIG_NUM_COOP_PRIORITIES
+
+#define FLASH_TIME_PER_WORD_US (CONFIG_EMDS_FLASH_TIME_TO_WRITE_ONE_WORD_US)
+
+#if CONFIG_BT
+#define FLASH_ENTRY_TIME_OVERHEAD_US (3*470)
+#else
+#define FLASH_ENTRY_TIME_OVERHEAD_US (3*14)
+#endif
+
+#if CONFIG_PARTIAL_ERASE
+#define FLASH_BASE_TIME_OVERHEAD_US 5000
+#else
+#define FLASH_BASE_TIME_OVERHEAD_US 100000
+#endif
+
+K_SEM_DEFINE(emds_sem, 0, 1);
+bool emds_store_in_progress = false;
+bool emds_ready = false;
+bool emds_initialized = false;
+int meta_data_size = 8;
+
+struct emds_storage {
+	struct emds_fs cf_emds;
+	const char *flash_dev_name;
+};
+
+sys_slist_t emds_dynamic_entries;
+static struct emds_storage default_emds_storage;
+
+static void emds_handler(void)
+{
+	k_sem_reset(&emds_sem);
+	k_sem_take(&emds_sem, K_FOREVER);
+
+	LOG_DBG("Emergency Data Storeage released");
+
+	k_sched_lock();
+	emds_store_in_progress = true;
+
+	STRUCT_SECTION_FOREACH(emds_entry, ch) {
+		ssize_t len = emds_flash_write(&default_emds_storage.cf_emds,
+					ch->id, ch->data, ch->len);
+		if (len == ch->len) {
+			LOG_DBG("Write static entry: %d, Data length: %d",
```

Comment:
Note: We should be very careful with logging inside this time sensetive portion of the code. If log is triggered during shutdown it will most certainly impact the execution time of the write procedure.